Adds helper methods to ease ADO.NET configuration

veikkoeeva opened the pull request:
The ADO.NET functionality should utilities such
as some other providers (e.g. AzureUtils).

veikkoeeva commented:
I'll see if there's an efficient way to add tests still. I intend this to be ready in a few hours. This is just to open a PR.

jdom commented on src/OrleansSQLUtils/Storage/AdoNetConfigurationExtensions.cs:
```diff
@@ -0,0 +1,89 @@
+            this ClusterConfiguration config,
+            string providerName = "AdoNetStorage",
+            string connectionString = null,
+            AdoNetSerializationFormat serializationFormat = AdoNetSerializationFormat.OrleansNativeBinary)
```
do we really want to have Orleans native as the default? I would really discourage this if there is some alternative that works fine. Orleans serializer is not version tolerant, which is a very bad thing for stuff persisted in durable storage. I think in newer providers (ie: azure blob storage) we even stopped providing Orleans binary serializer as an option.

veikkoeeva replied:
It can be changed. Would XML be a neutral choice?

In the future I personally would hope to remove this "string configuration" and provide actual libraries. The default could be what is included by default, such as XML and some Orleans fallback (de)serializer.

Non-version tolerant isn't good at least. The ADO.NET storage has an interceptor mechanism to handle it, but it's not accessible until DI or interface change since Orleans instantiates the class and the application code can't get a reference to it.

jdom replied:
You might know better, probably XML is common in DBs, but Json seems acceptable too. Since you and @shayhatsor own this area, I'm fine with whatever choice, it's just that I prefer not the non-version tolerant one, as it can be a source of issues that people get bitten with after they have code (and data) in prod.

veikkoeeva replied:
OK. I'll update withint the next 24 hours. Too tired now. :)

veikkoeeva commented:
There doesn't seem to be a way to fully test this functionality without doing much more work than I can afford currently. It looks like the other such methods aren't holistically tested either. However, I used the Storage sample to check the values are passed correlty and data is saved correctly.

jdom commented:
Seems fine for now. I think that once we move to having testing kits as we talked in https://github.com/dotnet/orleans/issues/2011 we should have a lot more silo and non-silo test coverage for every provider, and that will be an easy chance to use the extension methods too.

veikkoeeva commented:
@jdom I made XML as the default serialization choice. I also changed the namespace to the root in case there will be other helper methods besides storage related. I rebased out of habit also, hopefully it's OK.

attilah commented on src/OrleansSQLUtils/Storage/AdoNetConfigurationExtensions.cs:
```diff
@@ -0,0 +1,89 @@
+                throw new ArgumentNullException(nameof(providerName));
+            }
+
+            connectionString = GetConnectionString(connectionString, config);
```
IMO It is not a good practice to use an input parameter and modify it inside the body of a method. Same stands for serializationFormat below.

veikkoeeva replied:
Should it be changed? I believe this codebase (and plenty of code in the wild) modifies default value is if it is in illegate state. I don't see the `serializationFormat` being modified.

As you might be aware, this is practically a copy of `AzureConfigurationExtensions`, likely not the only place in this codebase and there's plenty of code elsewhere that does this too.

attilah replied:
Back in the day that's what I have been thought :-) Touching input parameters by modifying their values is a kind of code smell in managed code, but "where I came from" native code it can lead to unwanted, hard to track consequences, for example if you're modifying a pointer what you received as an input parameter.
I'd avoid at all times. Imagine that you're logging the values what you're calling a method with, then you see an error and you've no clue HOW that value got there since that is not what you passed in. If we "allow" this in the codebase it will make it hard to differentiate on a case by case basis that it is OK here to modify it, but there it can lead to hard to track errors.

jdom replied:
I understand that could be an issue in native code, but the reality is that not here in c# when we are dealing with references to immutable objects/structs.
But to add to it, IMO creating a new variable just so that we don't update the reference can cause more maintainability problems than just reusing it (as code could mistakenly use the input parameter instead of the sanitized one). Also for consistency with everywhere else we should let this in and if you feel strongly (I don't, so I'm open to change) we can open an issue and then update all our usages of input sanitization in one go, but not be prescriptive yet on the non-consistent approach until there's consensus.

attilah replied:
Ok.

veikkoeeva replied:
@attilah Here the comments tell the connection string should be given, but if not, for conveniene a global one will be acquired and used. Partial function application would probably work here, but maybe we should keep the configuration like the other storage configuration utilities?

jdom replied:
Since I think this is settled (at least temporarily, but future changes would go in a different issue anyway), I'll merge, since we want to pull this into 1.3.1
